File: frontend/src/api/prompts.ts
import { api } from './client'

export type PromptType =
  | 'generate'
  | 'review'
  | 'optimize'
  | 'wiki_analyze'
  | 'wiki_write'

export interface PromptTemplate {
  id: number
  name: string
  type: string
  content: string
  version: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface PromptCreate {
  name: string
  type: string
  content: string
  is_active?: boolean
}

export interface PromptUpdate {
  name?: string
  content?: string
  is_active?: boolean
}

export function listPrompts(type?: string) {
  const qs = type ? `?type=${encodeURIComponent(type)}` : ''
  return api<PromptTemplate[]>(`/api/prompts${qs}`)
}

export function createPrompt(body: PromptCreate) {
  return api<PromptTemplate>('/api/prompts', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function updatePrompt(id: number, body: PromptUpdate) {
  return api<PromptTemplate>(`/api/prompts/${id}`, {
    method: 'PUT',
    body: JSON.stringify(body),
  })
}

export const PROMPT_TYPE_OPTIONS: { value: PromptType; label: string }[] = [
  { value: 'generate', label: '生成 (generate)' },
  { value: 'review', label: '评审 (review)' },
  { value: 'optimize', label: '优化 (optimize)' },
  { value: 'wiki_analyze', label: 'Wiki 分析 (wiki_analyze)' },
  { value: 'wiki_write', label: 'Wiki 写入 (wiki_write)' },
]

File: backend/app/api/prompts.py
from datetime import datetime, timezone
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlmodel import Session, col, func, select

from app.db import get_session
from app.models.entities import PromptTemplate
from app.schemas.prompts import PromptCreate, PromptOut, PromptUpdate

router = APIRouter(prefix="/api/prompts", tags=["prompts"])


def _utcnow() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _next_version(session: Session, prompt_type: str) -> int:
    current = session.exec(
        select(func.max(PromptTemplate.version)).where(PromptTemplate.type == prompt_type)
    ).one()
    if current is None:
        return 1
    return int(current) + 1


@router.get("", response_model=List[PromptOut])
def list_prompts(
    type: Optional[str] = Query(default=None),
    session: Session = Depends(get_session),
) -> list[PromptTemplate]:
    stmt = select(PromptTemplate).order_by(col(PromptTemplate.id).desc())
    if type:
        stmt = stmt.where(PromptTemplate.type == type)
    return list(session.exec(stmt).all())


@router.post("", response_model=PromptOut)
def create_prompt(
    body: PromptCreate,
    session: Session = Depends(get_session),
) -> PromptTemplate:
    row = PromptTemplate(
        name=body.name,
        type=body.type,
        content=body.content,
        version=_next_version(session, body.type),
        is_active=body.is_active,
    )
    session.add(row)
    session.commit()
    session.refresh(row)
    return row


@router.get("/{prompt_id}", response_model=PromptOut)
def get_prompt(prompt_id: int, session: Session = Depends(get_session)) -> PromptTemplate:
    row = session.get(PromptTemplate, prompt_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Prompt not found")
    return row


@router.put("/{prompt_id}", response_model=PromptOut)
def update_prompt(
    prompt_id: int,
    body: PromptUpdate,
    session: Session = Depends(get_session),
) -> PromptTemplate:
    row = session.get(PromptTemplate, prompt_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Prompt not found")

    data = body.model_dump(exclude_unset=True)
    for key, value in data.items():
        setattr(row, key, value)
    row.updated_at = _utcnow()
    session.add(row)
    session.commit()
    session.refresh(row)
    return row
